lib/crypto/src/bigint.rs

```diff
@@ -0,0 +1,277 @@
+        <Uint<N> as Integer>::is_even(self).into()
+    }
+
+    fn is_zero(&self) -> bool {
+        <Uint<N> as Zero>::is_zero(self).into()
+    }
+
+    fn num_bits(&self) -> usize {
+        self.bits()
+    }
+
+    fn get_bit(&self, i: usize) -> bool {
+        self.bit(i).into()
+    }
+}
+
+impl<const N: usize> BitIteratorBE for Uint<N> {
+    fn bit_be_iter(&self) -> impl Iterator<Item = bool> {
+        self.as_words().iter().rev().flat_map(Word::bit_be_iter)
+    }
+}
+
+/// Parse a number from a string in a given radix.
+///
+/// I.e., convert string encoded integer `s` to base-`radix` number.
+#[must_use]
+pub const fn from_str_radix<const LIMBS: usize>(
+    s: &str,
+    radix: u32,
+) -> Uint<LIMBS> {
+    let bytes = s.as_bytes();
+
+    // The lowest order number is at the end of the string.
+    // Begin parsing from the last index of the string.
+    let mut index = bytes.len() - 1;
+
+    let mut uint = Uint::from_u32(0);
+    let mut order = Uint::from_u32(1);
+    let uint_radix = Uint::from_u32(radix);
+
+    loop {
+        // Try to parse a digit from utf-8 byte
+        let ch = parse_utf8_byte(bytes[index]);
+        let digit = match ch.to_digit(radix) {
+            None => {
+                panic!("invalid digit");
+            }
+            Some(digit) => Uint::from_u32(digit),
+        };
+
+        // Add a digit multiplied by order.
+        uint = add(&uint, &mul(&digit, &order));
+
+        // If we reached the beginning of the string, return the number.
+        if index == 0 {
+            return uint;
+        }
+
+        // Increase the order of magnitude.
+        order = mul(&uint_radix, &order);
+
+        // Move to the next digit.
+        index -= 1;
+    }
+}
+
+/// Multiply two numbers and panic on overflow.
+#[must_use]
+pub const fn mul<const LIMBS: usize>(
+    a: &Uint<LIMBS>,
+    b: &Uint<LIMBS>,
+) -> Uint<LIMBS> {
+    let (low, high) = a.mul_wide(b);
+    assert!(high.bits() == 0, "overflow on multiplication");
+    low
+}
+
+/// Add two numbers and panic on overflow.
+#[must_use]
+pub const fn add<const LIMBS: usize>(
+    a: &Uint<LIMBS>,
+    b: &Uint<LIMBS>,
+) -> Uint<LIMBS> {
+    let (low, carry) = a.adc(b, Limb::ZERO);
+    assert!(carry.0 == 0, "overflow on addition");
+    low
+}
+
+/// Parse a single UTF-8 byte.
+pub(crate) const fn parse_utf8_byte(byte: u8) -> char {
+    match byte {
+        0x00..=0x7F => byte as char,
+        _ => panic!("non-ASCII character found"),
+    }
+}
+
+/// This macro converts a string base-10 number to a big integer.
+#[macro_export]
+macro_rules! from_num {
+    ($num:literal) => {
+        $crate::bigint::from_str_radix($num, 10)
+    };
+}
+
+/// This macro converts a string hex number to a big integer.
+#[macro_export]
+macro_rules! from_hex {
+    ($num:literal) => {
+        $crate::bigint::crypto_bigint::Uint::from_be_hex($num)
+    };
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn convert_from_str_radix() {
```

Comment:
I would say that these are perfect tests for fuzzing 😅

Reply:
I think you're right. We can try to add them, but for relatively small numbers probably (up to u128) , to be able test them with convenience

Reply:
Wdt to add it within Poseidon hash pr? And live it like that for a while